### PR TITLE
TCOMP-1771 : record builder optim

### PR DESCRIPTION
**Pull request by clesaec**
### Requirements
JIRA => https://jira.talendforge.org/browse/TCOMP-1771
* Any code change adding any logic **MUST** be tested through a unit test executed with the default build
* Any API addition **MUST** be done with a documentation update if relevant 

### Why this PR is needed?
basic optimization 
### What does this PR adds (design/code thoughts)?
no design added, code optim & tested.

### CODE REVIEW

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java [hunk 1]
```diff
@@ -115,25 +115,37 @@ public BuilderImpl(final Schema providedSchema) {
...
                 throw new IllegalArgumentException(
                         "No entry '" + name + "' expected in provided schema: " + entryIndex.keySet());
             }
-            if (entry.getType() != type) {
-                throw new IllegalArgumentException(
-                        "Entry '" + name + "' expected to be a " + entry.getType() + ", got a " + type);
+            return entry;
+        }
+
+        private Schema.Entry findOrBuildEntry(final String name, final Schema.Type type, boolean nullable) {
```
Comment (undx) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:131]: `nullable` should be f`final`.

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java [hunk 2]
```diff
@@ -115,25 +115,37 @@ public BuilderImpl(final Schema providedSchema) {
...
         }
 
-        private void validateTypeAgainstProvidedSchema(final String name, final Schema.Type type, final Object value) {
-            if (providedSchema == null) {
-                return;
-            }
-            if (entryIndex == null) {
-                entryIndex = providedSchema.getEntries().stream().collect(toMap(Schema.Entry::getName, identity()));
+        private Schema.Entry findMandatoryEntry(final String name) {
+            assert this.providedSchema != null;
```
Comment (undx) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:119]: why using `assert` here?

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java [hunk 3]
```diff
@@ -115,25 +115,37 @@ public BuilderImpl(final Schema providedSchema) {
             this.providedSchema = providedSchema;
         }
 
-        private void validateTypeAgainstProvidedSchema(final String name, final Schema.Type type, final Object value) {
-            if (providedSchema == null) {
-                return;
-            }
-            if (entryIndex == null) {
-                entryIndex = providedSchema.getEntries().stream().collect(toMap(Schema.Entry::getName, identity()));
+        private Schema.Entry findMandatoryEntry(final String name) {
```
Comment (undx) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:118]: function name: remove mandatory or rename it ...

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java [hunk 4]
```diff
@@ -115,25 +115,37 @@ public BuilderImpl(final Schema providedSchema) {
...
+        }
+
+        private Schema.Entry findOrBuildEntry(final String name, final Schema.Type type, boolean nullable) {
+            if (providedSchema == null) {
+                return new SchemaImpl.EntryImpl.BuilderImpl().withName(name).withType(type).withNullable(nullable).build();
+            }
+            return this.findMandatoryEntry(name);
+        }
+
+        private Schema.Entry validateTypeAgainstProvidedSchema(final String name, final Schema.Type type, final Object value) {
```
Comment (undx) [inline component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:138]: Here, 
- the function name do not match the goal of the function and provided type is useless.
- the function has a side effect by creating a new entry and it doesn't match initial goal.

**Review by undx — COMMENTED** [review]
see my comments

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 2 above)
Comment (clesaec) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:119]: this function is private and must be called only when providedSchema exists. assert is normally not tested on running code, just in unit test (until we used -ea option)

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 3 above)
Comment (clesaec) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java]: renamed findExistingEntry, better ? why mandatory was not good ?

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 1 above)
Comment (clesaec) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java]: done

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 4 above)
Comment (clesaec) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:138]: - type checking added (don't know if it's usefull, assertType do also this control)
- this method never create entry.

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 3 above)
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java]: sorry a matter of taste... ;-)

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 2 above)
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:119]: still think that its place in in tests not here...

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 4 above)
Comment (undx) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java:138]: you're right on second point ;-)

File: component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java (hunk 2 above)
Comment (clesaec) [reply component-runtime-impl/src/main/java/org/talend/sdk/component/runtime/record/RecordImpl.java]: assert removed

**Review by undx — APPROVED** [review]
LGTM

**Pull request merged**